scheduler/workflow.go

```diff
@@ -335,26 +336,38 @@ func (s *schedulerWorkflow) StateString() string {
 }
 
 func (s *schedulerWorkflow) workJobs(prs []*processNode, pus []*publishNode, t *task, pj job) {
+	// create a waitgroup track pending goroutines submitting jobs
+	wg := &sync.WaitGroup{}
 	for _, pr := range prs {
-		j := newProcessJob(pj, pr.Name(), pr.Version(), pr.InboundContentType, pr.config.Table(), t.metricsManager, t.id)
-		errors := t.manager.Work(j).Promise().Await()
-		if len(errors) != 0 {
-			t.failedRuns++
-			t.lastFailureTime = t.lastFireTime
-			t.lastFailureMessage = errors[len(errors)-1].Error()
-			return
-		}
+		go func() {
+			wg.Add(1)
+			defer wg.Done()
+			j := newProcessJob(pj, pr.Name(), pr.Version(), pr.InboundContentType, pr.config.Table(), t.metricsManager, t.id)
+			errors := t.manager.Work(j).Promise().Await()
+			if len(errors) != 0 {
+				t.failedRuns++
```

Comment:
Here is more detail to think about:

Only the task itself reaches workJob and by design the function at the top only allows one root goroutine to ever be running workJob:
1. spin() blocks while executing - https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L256
2. spin() calls fire() - https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L273
4. fire() call starts the workflow - https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L256
5. the workflow calls workJobs - https://github.com/intelsdi-x/snap/blob/master/scheduler/workflow.go#L326

The risk of mutation on task is on the properties that this flow accesses which are:
1. deadlineDuration (mutable and thread safe)
2. metricsManager (immutable)
3. *schedulerWorkflow (immutable - workflows cannot be altered)
4. task id (immutable)
5. failedRuns, lastFailureTime, lastFailureMessage (mostly thread safe, probably need a mutex on these to ensure updates to all three are atomic). Two ops writing at the same time does not cause problems. But error messages would be mixed. Acquiring `t's` lock would be an easy way to synchronize this.

The only issue here is that by acquiring t's lock we are preventing tasks from being stopped until the entire workflow is worked (not submitted) since the Work(j).Promise().Await() is blocking. This is what bothers me since calls to change state on task are not async. A long running workflow would block something that should work. This is why I would prefer not to mix the state change mutex `t.Lock()` with a possible mutex around internal properties of `t` that we need to synchronize on.

To add flavor to the above, workJobs does not submit all of the work at once. Each iteration on a process node has to complete before calling the child node. So we would not be locking `t` and submitting all the jobs. We would be locking `t` for the duration of all jobs where the parent node has a child.

We would basically move the blocking operation we had on submission up to state changes on task.